src/distributed_graph_generator.cc

```diff
@@ -0,0 +1,608 @@
+		for(auto d : last_writer_cmd->get_dependents()) {
+			// Only consider true dependencies
+			if(d.kind != dependency_kind::true_dep) continue;
+
+			const auto cmd = d.node;
+
+			// We might have already generated new commands within the same task that also depend on this; in that case, skip it
+			if(isa<task_command>(cmd) && static_cast<task_command*>(cmd)->get_tid() == tid) continue;
+
+			// So far we don't know whether the dependent actually intersects with the subrange we're writing
+			if(const auto command_reads_it = m_command_buffer_reads.find(cmd->get_cid()); command_reads_it != m_command_buffer_reads.end()) {
+				const auto& command_reads = command_reads_it->second;
+				// The task might be a dependent because of another buffer
+				if(const auto buffer_reads_it = command_reads.find(bid); buffer_reads_it != command_reads.end()) {
+					if(!GridRegion<3>::intersect(write_req, buffer_reads_it->second).empty()) {
+						has_successors = true;
+						m_cdag.add_dependency(write_cmd, cmd, dependency_kind::anti_dep, dependency_origin::dataflow);
+					}
+				}
+			}
+		}
+
+		// In some cases (horizons, master node host task, weird discard_* constructs...)
+		// the last writer might not have any successors. Just add the anti-dependency onto the writer itself then.
+		if(!has_successors) { m_cdag.add_dependency(write_cmd, last_writer_cmd, dependency_kind::anti_dep, dependency_origin::dataflow); }
+	}
+}
+
+void distributed_graph_generator::process_task_side_effect_requirements(const task& tsk) {
+	const task_id tid = tsk.get_id();
+	if(tsk.get_side_effect_map().empty()) return; // skip the loop in the common case
+	if(m_cdag.task_command_count(tid) == 0) return;
+
+	for(const auto cmd : m_cdag.task_commands(tid)) {
+		for(const auto& side_effect : tsk.get_side_effect_map()) {
+			const auto [hoid, order] = side_effect;
+			if(const auto last_effect = m_host_object_last_effects.find(hoid); last_effect != m_host_object_last_effects.end()) {
+				// TODO once we have different side_effect_orders, their interaction will determine the dependency kind
+				m_cdag.add_dependency(cmd, m_cdag.get(last_effect->second), dependency_kind::true_dep, dependency_origin::dataflow);
+			}
+
+			// Simplification: If there are multiple chunks per node, we generate true-dependencies between them in an arbitrary order, when all we really
+			// need is mutual exclusion (i.e. a bi-directional pseudo-dependency).
+			m_host_object_last_effects.insert_or_assign(hoid, cmd->get_cid());
+		}
+	}
+}
+
+void distributed_graph_generator::set_epoch_for_new_commands(const abstract_command* const epoch_or_horizon) {
+	// both an explicit epoch command and an applied horizon can be effective epochs
+	assert(isa<epoch_command>(epoch_or_horizon) || isa<horizon_command>(epoch_or_horizon));
+
+	for(auto& [bid, bs] : m_buffer_states) {
+		// TODO this could be optimized to something like cdag.apply_horizon(node_id, horizon_cmd) with much fewer internal operations
```

Comment:
Is this comment still relevant? (I dimly recall adding it long ago, but the underlying mechanisms were very different)

Reply:
Yeah I don't think it's relevant anymore, removed.